src/contexts/ActivityContext.js

```diff
@@ -0,0 +1,238 @@
+import React from 'react'
+import PropTypes from 'prop-types'
+import StoredList from '../StoredList'
+import { network } from '../environment'
+import { EthereumAddressType } from '../prop-types'
+
+const ActivityContext = React.createContext()
+
+const activityStatusTypes = {
+  CONFIRMED: 'CONFIRMED',
+  PENDING: 'PENDING',
+  FAILED: 'FAILED',
+  TIMED_OUT: 'TIMED_OUT',
+}
+
+const activityTypes = {
+  TRANSACTION: 'TRANSACTION',
+}
+
+const TEN_MINUTES = 1000 * 60 * 10
+
+let storedList
+
+// Provides easy access to the user activities list
+class ActivityProvider extends React.Component {
+  static propTypes = {
+    account: EthereumAddressType, // Current wallet
+    children: PropTypes.node,
+    daoDomain: PropTypes.string, // domain of current DAO
+  }
+  state = {
+    // activities of all accounts
+    activities: [],
+  }
+  static defaultProps = {
+    account: '',
+  }
+
+  componentDidUpdate(prevProps) {
+    const { daoDomain } = this.props
+
+    if (daoDomain !== prevProps.daoDomain) {
+      const storageKey = `activity:${network.type}:${daoDomain}`
+      storedList = new StoredList(storageKey)
+
+      // eslint-disable-next-line react/no-did-update-set-state
+      this.setState({ activities: storedList.getItems() })
+    }
+  }
+
+  componentWillMount() {
+    this._checkInterval = setInterval(this.checkForTimedOut, 1000 * 30)
+  }
+  componentWillUnmount() {
+    clearInterval(this._checkInterval)
+  }
+
+  checkForTimedOut = () => {
+    const now = Date.now()
+    this.state.activities.forEach(activity => {
+      const timeDelta = now - activity.createdAt
+      if (
+        timeDelta > TEN_MINUTES &&
+        activity.status === activityStatusTypes.PENDING
+      ) {
+        // Set pending items to timed out after 10 minutes
+        this.setActivityTimedOut(activity.transactionHash)
+      }
+    })
+  }
+
+  add = activity => {
+    this.setState({
+      activities: storedList.add(activity),
+    })
+  }
+
+  addTransactionActivity = ({
+    transactionHash = '',
+    from = '',
+    targetAppProxyAddress = '',
+    forwarder = '',
+    forwarderProxyAddress = '',
+    description = '',
+  } = {}) => {
+    const newActivity = {
+      createdAt: Date.now(),
+      type: activityTypes.TRANSACTION,
+      status: activityStatusTypes.PENDING,
+      read: false,
+      transactionHash,
+      // account address from which the transaction was created
+      from,
+      targetAppProxyAddress,
+      forwarderProxyAddress,
+      forwarder,
+      description,
+    }
+
+    const updatedActivities = storedList.add(newActivity)
+
+    this.setState({ activities: updatedActivities })
+  }
+
+  currentAccountPredicate = ({ from }) =>
```

Comment:
Wondering if it would be easier if we made the storage key unique to the current account too, so we don't need to do this filtering?

Reply:
Most likely, in fact that was how it was at an earlier revision. I think this was so that I could still "catch" updates to activities in the situation where a user switches account/wallet before receiving confirmation. However, now that I've implemented `refreshPendingActivities` we should be good to safely miss the updates originating from the web3 receipt callback.